### ISSUE TITLE
When updating state of changed entity related entity is not updated with new relation.

### ISSUE THREAD

**Issue by evolve2k** [dixieio/redux-json-api#62]
I'm following the API approach for createEntity.

I am updating an entity which is a child of a related entity, to illustrate I'll refer to task and TaskList from the official example as it is adequate currently to illustrate my point.

When createEntity is dispatched it is correctly updating the task to my backend. It is also updating state to include the new task BUT TaskList is not having it's state updated to reflect that it now has this additional task related to it and hence in my UI the new task is not showing as only tasks related to my TaskList show.

Should I be dispatching something further to ensure the TaskList updates to include the new task or should all related entities be updating following the  dispatch(createEntity(entity)) as below.

_Per: https://github.com/dixieio/redux-json-api/blob/master/docs/apis/createEntity.md_

```
class CreateTask extends Component {
  handleSubmit() {
    const { dispatch } = this.props;
    const entity = {
      type: 'tasks',
      attributes: {
        task: 'New task name'
      },
      relationships: {
        taskList: {
          data: {
            id: '1',
            type: 'taskLists'   <==== Does this also have its relations updated from the dispatch?
          }
        }
      }
    }

    dispatch(createEntity(entity));
  }
```

**Comment by egeriis** [comment]
Hi @evolve2k,

Thank you for reaching out.

I need a bit more information, to answer your question. The update of reverse relationships are dependent on the relationship existing in both ends. I.e. the _taskList_ must have a `task` or `tasks` relationships key for it to be updated, when a _task_ has been updated. Likewise the _task_ must have the reverse relationship.

So what is your payload, how is this handled by _redux-json-api_ and what is expected outcome?

**Comment by evolve2k** [comment]
I understand mostly what you're asking and I hope I can form an answer that makes sense.
My backend API has relationships between my equivilent of TaskList and Task. Each TaskList has many Tasks and each Task has one TaskList. 
(It's a Rails 4.2 backend using the 'react_on_rails' gem for what it's worth).

When I add an item (and include the foreign key for its parent TaskList) I can see that the correct request has been sent across to my backend and my redux state reports as API_CREATED but the actual state of the app is not updated as I want.

Simply performing the dispatch(createEntity(entity)) as above is not returning an updated state for TaskList and as such the parent TaskList does not include the new task. 
The task is listed in the redux dev tools as existing but since my UI shows a TaskList and all of it's children, then because this task is not yet a child then nothing is displaying.

If I refresh the page the app performs an API_READ and this pulls in an updated TaskList that now includes the related task, and so the new task displays then but obviously having it work just on page reload is not quite the point.

Which path should I be heading down...
1. Have I just got my relationship syntax wrong and should be expecting redux-json-api to be updating TaskList to include by new task by merit of including relationships in the create task request; OR
2. Should I be sending a second dispatch to explicitly add the task to the TaskList within the store; 
3. Should I be triggering some kind of subscribe or forced API_READ after doing a dispatch so that the new cross relation is pulled back from the API after the initial request; OR
4. What should I be doing?

Thanks

**Comment by evolve2k** [comment]
Wow, somehow I was *that close*, got this working.  For anyone following my footsteps it seems that path 1 is closest to being correct. I just had a bit of syntax wrong and once it was working the new "tasks" started to appear in the view.

Checking the state, the task is added (as I can confirm in the Redux Dev tools with API_CREATED), the task does not appear in the state of the related entity, but now that it appears on screen where I expected it and that it is available to be updated/deleted then this is the real functionality I needed.

I'll test a few more things then will close this in a day or so unless I find I've missed something.

Thanks for the friendly prompt support!

**Comment by evolve2k** [comment]
This is working in one part of my app but not another. I still don't understand it correctly.
Seems like a bit of magic how the parent related entity gets updated by redux-json-api in the store. Can someone shed more light and give me a explaination of how this part works>

**Comment by evolve2k** [comment]
Just to document my progress, it turns out that the very popular ActiveModelSerializer ruby gem created by the creators of JSONAPI actually doesn't support the full spec!!! 

**TLDR; If you're using JSON API with Rails get off of ActiveModelSerializer and shift to JSONAPI_resources.**

### Why?

> .. the problem is that ActiveModelSerializer doesn't actually provide us with a way of adapting inputs say if we try to create another 'post' using JSON API it will not understand what is happening. So if you are only concerned with serializing your data and getting it out ActiveModelSerializer has got your back. But if we want to support having like a create action it's not going to understand how to interpret the JSON API formatted data. At this point Id recommend reaching for something like **JSONAPI_resources**. This provides more of a framework actually for creating an API with create and update actions that follow the JSON API spec.

Reference: JSON API with Rails 5 (https://youtu.be/EpzIhQBXUQk?t=9m10s)

Im now migrating across to using JSONAPI_resources on my backend API, I'm sure this will raise new questions.

**Comment by egeriis** [comment]
Relationships are a bit magic with regards to the JSON API v1 spec as they don't specify their inverse relation. Thus the implementation in _redux-json-api_ assumes some things, that may provide needed functionality for most cases but not in all:

1. When _redux-json-api_ reads and parses a resource, it will look through relationships of that resource
1. The relationship of a resource has a few pieces of information that it uses: resource type, relationship name, foreign type, foreign id
1. _redux-json-api_ will look for resources of given _foreign type_ with specified _foreign id_
1. If a foreign resource is found, _redux-json-api_ will look for a relationship _named_ after the resource type of the original resource (plural or singular) and update it accordingly

There are more steps, e.g. determining if the inverse relationship is has-many or has-one. But above should explain the flow to more detail.

--

Regarding the specific issue you're having, I would need the following information:

1. Full Redux state from `api` namespace before request
1. Response body from your POST request
1. Full Redux state form `api` namespace after request

### Example

#### State before request

```json
{
  "tasks": {
    "data": [
      { 
        "type": "tasks",
        "id": 1,
        "attributes": {
          "title": "Lorem ipsum"
        },
        "relationships": {
          "taskList": {
            "data": {
              "type": "taskLists",
              "id": 1
            }
          }
        }
      }
    ]
  },
  "taskLists": {
    "data": [
      {
        "type": "taskLists",
        "id": 1,
        "attributes": {
          "title": "Task list 1"
        },
        "relationships": {
          "tasks": {
            "data": [
              {
                "type": "tasks",
                "id": 1
              }
            ]
          }
        }
      }
    ]
  }
}
```

#### Request response

```json
{
  "data": {
    "type": "tasks",
    "id": 2,
    "attributes": {
      "title": "New task"
    },
    "relationships": {
      "taskList": {
        "data": {
          "type": "taskLists",
          "id": 1
        }
      }
    }
  }
}
```

#### State after request

1. New task added to `tasks`
1. Inverse relationship for `taskList` relation on new task is registered

```json
{
  "tasks": {
    "data": [
      { 
        "type": "tasks",
        "id": 1,
        "attributes": {
          "title": "Lorem ipsum"
        },
        "relationships": {
          "taskList": {
            "data": {
              "type": "taskLists",
              "id": 1
            }
          }
        }
      },
      {
        "type": "tasks",
        "id": 2,
        "attributes": {
          "title": "New task"
        },
        "relationships": {
          "taskList": {
            "data": {
              "type": "taskLists",
              "id": 1
            }
          }
        }
      }
    ]
  },
  "taskLists": {
    "data": [
      {
        "type": "taskLists",
        "id": 1,
        "attributes": {
          "title": "Task list 1"
        },
        "relationships": {
          "tasks": {
            "data": [
              {
                "type": "tasks",
                "id": 1
              },
              {
                "type": "tasks",
                "id": 2
              }
            ]
          }
        }
      }
    ]
  }
}
```

**Comment by evolve2k** [comment]
Thanks for offering to help @egeriis.
It's taken me a while to get my head around how I can best share this with you as the json, is very long and it's been a bit of messing around to get it shorter.

All I'm needing to achieve is basic CRUD actions between my redux app and my rails backend. I keep feeling like it's one step forward and two steps back, none-the-less, I hope with a little help from you I can get to the bottom of this.

1. Full Redux state from api namespace before request: 
Here is the request, it's in a gist as it's so long
https://gist.github.com/evolve2k/82618e06d3ca0c4360a092b7d1482423

2. Response body from your POST request
I'm having a few different issues.

Issue A. The first is that income items are showing correctly but zen accounts and their children  expense items are not showing. Why are income items showing but zen accounts  are not?

Issue B. I can see and interact with income-items but when I attempt to delete them I'm getting this console error. 

When attempting Delete:
`utils.js?5810:13 DELETE http://localhost:3000/api/v1/incomeItems/138 404 (Not Found)`

When attempting Update:
utils.js?5810:13 PATCH http://localhost:3000/api/v1/incomeItems/138 404 (Not Found)

I know what's happening here, my API is only able to consume uri's that are caterpillar case. When I changed parts of the code to list 'income-items' rather than 'incomeItems' then the reading of the records breaks and all the incomes stopped showing in my react output. 

Adding income items is working currently and i think that's because I can control the exact syntax of the type within my add function so that it's matching the API.

Probably not much use providing API json after the request just yet as the issue is very clear with relation to the delete and update and the read doesn't have an after state. Happy to provide anything else you need.

**Comment by evolve2k** [comment]
Note, on the server side (Rails 4.2 using the JSON_Resources gem), I've reconfigured the API to serve up camelCase as keys within the JSON. 

JSON_Resources defaults to _caterpillar-case_ or as they call it _dasherized_ but allows you to reconfigure the key type. It seems that the uri format is fixed as _caterpillar-case_.

If redux-json-api supported caterpillar case I think this would just work end-to-end but with this quirk it looks like I need to specify my JSON to match your preference for camelCase and then work out a way to hack the uri's back to caterpilla case.

**Comment by egeriis** [comment]
Let me get back asap. But regarding resource URLs I reckon that they need to be camel case according to spec.

**Comment by evolve2k** [comment]
I've made progress with this, I must be loosing my marbles but finally worked out that the library on the other side (jsonapi-resources) does let me specify the path case type. Eg I can now support camelCase via the API endpoint.

Here's the next thing. Now it's all wired up. The Read and Add methods are working correctly but the update and delete are having strange issues. Let me explain.

When I update an entry, a new record is added at the bottom of the table that is a duplicate of the entry Im editing. What I've worked out is happening is that within redux state. 

api.expenseItems is staying with say the original 2 expense items.
but api.zenAccounts.relationships.expenseItems is increasing to include a new duplicate of the expense Im editing, each time I edit a field on that expense item (my code triggers an onBlur that leads to a dispatch update any time a field is changed on an expense row).

I think what is happening is my server response for the update is including the relationship data and redux-json-api is then adding a new record against the related parent entity which happens to be a duplicate of an existing entity and my redux table of expense items just then showing both as duplicates. Could it be that the cross relationship code is dumb so that it's not checking a relation already exists before adding the reverse relationship. Am I therefore right in thinking your expected fix is to get the API endpoint to just return a response that doesn't include the relation?

**Comment by evolve2k** [comment]
I've just changed the setting `config.always_include_to_one_linkage_data = true` on the ruby gem jsonapi-resources from true to false, and when testing update this has fixed the issue. Which is telling me that yes indeed redux-json-api currently does not want the relationship data included for updates. I'm currently working with other library to get fine grain ability to exclude the relationships from the response based on action. I'm not sure on your end it would be easy for me to mess with the response before it's processed in.

I've re-read over the spec, im not sure if it's clear but Im imagining that it means that if a relationship is included that on your end you should probably just recognise there is an existing connection and not create a duplicate item. Im interested in your thoughts (but I think my quickest fix is just to try to change the server response for now).

http://jsonapi.org/format/#crud-updating-resource-relationships

**Comment by evolve2k** [comment]
Hacked this into the server side and its removed the relations that we don't want to be included in the response to an update action. That's got the update working.
https://github.com/cerebris/jsonapi-resources/issues/871

**Comment by evolve2k** [comment]
Ok another big session pairing on this and I think there is a real issue somewhere with how the library is currently handling relationships.

For explanation lets call the main item the 'entity' and the related entity the 'parent'.

In simple terms it seems that if the server response includes relationships in the response that redux-json-api is:

1) Updating the entity as desired
2) Updating the parent, not detecting that the entity is already listed on the parent and then adding a new object under the parent relations so that our entity is now in the parents relations twice.
3) Displaying duplicated entities in view as a result.

This issue affects each of the CRUD actions in a different way creating a catch 22 when working with relations with redux-json-api.

After much testing I worked out that this issue does not occur when the server response does not include related entities in the response.

1. For create actions we want the related entity included in the response, once this has occured then redux-json-api handles everything nicely. There is no existing relation to the parent so creating the entity is the store and its reverse relation in the parent is the right thing to do.

2 For update actions if relationships are included in the server response the above reported duplication is occuring. On the surface a quick hack is to have the server respond with no relations but this causes flow on issues whereby now these relations have been stripped from the state, in my case breaking functionality around child relations to the entity updated. 

3. For DELETE actions technically the server should include no response, and redux-json-api should just respond to accept the 204. Delete actions are triggering correctly from redux-json-api with a valid request and triggering a correct 204 response from the server. What is not working though is that removeEntityFromState (within statemutation.js line 89) is not having updateReverseRelationship correct trigger and the entity is staying related to the parent entity and so for me where entities are listed within my view under the parent, the entity is not disapearing from the screen.

I really appreciate the help you guys are giving but I want to pleed a little further for one of you to have a dig around on this as since starting with this library almost a month ago now it's really the relationship code I've been struggling with.  I think this code might need a review/refactor and a couple more tests which should hopefully help unpack what's going on here.

If you can help me get this last bit sorted, that would be really greatly appreciated.

My use case is that of a parent entity that contains child entities and those entities contain their own child entities. If you could test a case with nested grandparent, child, grandchildren structure I think these issues should become readily apparent.

It might be as simple as just resolving the duplicate issue on the update but I just want to flag where to look and the depth of this issue.

@geriis I hope you can help.

**Comment by egeriis** [comment]
@evolve2k Thank you so much for digging into this! 

We would love to see some example response body, which will trigger these errors. As I understand, you're experiencing two issues:

1. Duplicate reverse relationships created upon resource patching
1. Deleting a resource does not remove it's reference from relationships

I would like to see the response body for the PATCH request and for you to confirm that these are the two issues you're experiencing.

Then we can go ahead and create a test to confirm the bug.

**Comment by evolve2k** [comment]
> I would like to see the response body for the PATCH request and for you to confirm that these are the two issues you're experiencing.

Yes that is correct, we are having issues on UPDATE and DELETE.

I've had to remove random hacks to get this into a clean state to share with you and sadly for debugging the DELETE is not failing right now as expected (darn computers, why work with them!). Lets work on the UPDATE and I'm sure the DELETE will fall over for me later and I can include the responses when I work that out, just know Im sure there is issues with DELETE too but just can't reproduce as of today.

I've played around to create responses as clearly as I can.
When no relationships are included everything works as expected, however currently the system is creating duplicates on update when relationship data is included as part of the request.

See this gist I've created.

It includes a working response (when no relationship data is included), the failing request (when relationship data IS INCLUDED) and also a screenshot of my redux state that shows the duplicate entry that is created (whereas the system should recognise an related entity is present).

Hope this gives you enough to work with.

https://gist.github.com/evolve2k/91e0f661d6d7399b04df756d101efc47

**Comment by egeriis** [comment]
@evolve2k Concerning the DELETE request, we do have a test which passes for the case you're experiencing. Maybe it's an edge case, which is hard to reproduce. I hate those, and I can hear you do too 😞

Do you mind sharing the entire _api_ namespace of state, before the PATCH request? Then I can quickly run a test. I'll be on top of this.

**Comment by evolve2k** [comment]
@egeriis thanks for your help!

Ok I obtained it this way, I made the error causing change (doing an UPDATE) and then went to the Chrome Redux toolbar > state > api and then clicked export below and downloaded the file 'state.json'.

This is exactly what its returning unmodified, hope that's what you need, let me know if you need anything else.

https://gist.github.com/evolve2k/a987fdb8152e7cb1cb457bacc34eabcd

**Comment by egeriis** [comment]
@evolve2k Wow, that's very extensive. Looks like it contains all actions and changes, that will not be too easy for me to use.

Do you have React dev tools? If so, could you go to the React dev tools, select `<Root>` and paste this into the console?

`JSON.stringify($r.props.store.getState().api)`

This should create a JSON output of what I'm looking for.

**Comment by evolve2k** [comment]
@egeriis Hmm yes it is rather long, ok I've done it the way you asked, see here, it looked easier to navigate in the console output actually, so I've added a screenshot of the console also.

Thanks again!

https://gist.github.com/evolve2k/e0f3725c32fc3508058503ea75776f6e

**Comment by evolve2k** [comment]
Oh crap just realised I just added all that from a clean state, eg before attempting the action that causes the double up. Just a minute, I'll repost with the correct state.

**Comment by evolve2k** [comment]
![myapp](https://cloud.githubusercontent.com/assets/114459/19599498/8e6d5b18-97e8-11e6-81bf-f1bb855688f4.jpg)
![myapp](https://cloud.githubusercontent.com/assets/114459/19599515/9d3dc592-97e8-11e6-8671-a55ac6004840.jpg)

**Comment by evolve2k** [comment]
Revised, using the store _after the UPDATE action is taken_.

https://gist.github.com/evolve2k/0dafc4469f6b7c1e2e3b6618656e37a3

**Comment by egeriis** [comment]
@evolve2k Which one is _before_ the PATCH request? You wrote:

> Using the store _after the UPDATE_ action is taken.

But I need the state before. 

Context: I am going to take "before state", paste it into a unit test, run the PATCH against it to enable testing and narrowing down the error.

**Comment by evolve2k** [comment]
@egeriis I just relabelled them, it's the first of the two.
Thx :)

**Comment by egeriis** [comment]
@evolve2k I left a comment on your gist: https://gist.github.com/evolve2k/e0f3725c32fc3508058503ea75776f6e

**Comment by evolve2k** [comment]
@egeriis no idea why those spaces were there, here is a version without those extra spaces. 
(It's in the clean state requested before any UPDATE is attempted).

Also I stripped back the clients record to have much less data before generating the JSON so now the file is much shorter. Should be much easier to read now.

https://gist.github.com/evolve2k/fdc2e4dd4cb3239ad1f82f96ab439dcd

**Comment by moxvallix** [comment]
@egeriis hey mate, im sure you're pretty busy but was hoping to poke this one along.
If from what I've provided its still hard to work out what's going on, I was thinking maybe we could pair on it. Happy to show my code if you're willing to have a poke around, thinking that might just be the quickest way if the smaller request I provided above is still not quite giving you any clarity.

**Comment by evolve2k** [comment]
[Edit: laptops having issues, jumped on my friends computer, just realised
I posted this as them accidently as they were logged into GH, deleting and
reposting as me! This is @evolve2k]

@egeriis hey mate, im sure you're pretty busy but was hoping to poke this
one along.
If from what I've provided its still hard to work out what's going on, I
was thinking maybe we could pair on it. Happy to show my code if you're
willing to have a poke around, thinking that might just be the quickest way
if the smaller request I provided above is still not quite giving you any
clarity.

**Comment by egeriis** [comment]
@evolve2k You're right, I am. As the CTO I am very on/off :) But I also just had a brief vacation in Hamburg. I'll get back during the weekend or Monday at latest. Hope that works for you.

**Comment by evolve2k** [comment]
Hey @egeriis yes I'd love to pair it out if that's possible. Hope you had a great holiday.
I'll send you a message on twitter and if you're happy to follow me back maybe we can DM and make a time. (Suggesting to do it that way to increase privacy & cut down noise in the channel here)

**Comment by evolve2k** [comment]
Following our discussion today. 
State and patch json.
https://gist.github.com/evolve2k/a4e8e60283d698685821fa8ea0fcbf73

**Comment by evolve2k** [comment]
Just confirming that #93 has indeed fixed my problem. 
Thanks again for all your help @egeriis.
